forge-ai/src/main/java/forge/ai/simulation/PossibleTargetSelector.java

```diff
@@ -203,8 +202,8 @@ public boolean selectTargetsByIndex(int targetIndex) {
     }
 
     public boolean selectTargets(Targets targets) {
-        if (targets.originalTargetCount != validTargets.size() || targets.targetingSaIndex != targetingSaIndex) {
-            System.err.println("Expected: " + validTargets.size() + " " + targetingSaIndex + " got: " + targets.originalTargetCount + " " + targets.targetingSaIndex);
+        if (targets.targetingSaIndex != targetingSaIndex) {
```

Comment:
I have some doubts this addresses the real cause...
Instead it looks more to me as if `originalTargetCount` merely is ambigiuously named because the only time it gets initialized is by the constructor which also uses `validTargets.size()`

Maybe something else messes up the count like the TargetUnique$?

Reply:
Well, this is the *real* cause - i.e. the plan fails because of this condition failing and the return false here.

This check is meant to be an assert - but clearly, it's not doing the right thing as without it, this case works correctly and with it, it errors out.

Reply:
But you're right, both are based on `validTargets.size()` - so it's not a mismatch between two different concepts, it is probably something like `TargetUnique$`.

Reply:
Ok so after debugging it looks to me like a lot of uneccessary/illegal simulations are being run because the selectors are all reset at once, which means it can't account for dependencies that should be enforced between their targets...

I've tried your test case ingame with the removed assert part (modified against empty opponents board for now) and can still run into a failure:
> Incremental Growth (937) - [Couldn't add to stack, failed to target] - Put a +1/+1 counter on target creature, two +1/+1 counters on another target creature, and three +1/+1 counters on a third target creature.[[Grizzly Bears (993)], [Runeclaw Bear (995)], []]

To address that root cause I've made this modification:
![grafik](https://user-images.githubusercontent.com/8506892/210091189-19a99cba-149e-48fc-84a3-a40751258181.png)
where ``softreset`` is ``reset()`` but still retains the current ``targetIndex``

This also passes your test.

The number of invalid simulations also gets reduced somewhat:

Before:
```
0: [879] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Grizzly Bears (3)], [Grizzly Bears (3)])
0: [954] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Grizzly Bears (3)], [Forest Bear (4)])
0: [954] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Grizzly Bears (3)], [Runeclaw Bear (5)])
0: [929] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Forest Bear (4)], [Forest Bear (4)])
0: [1004] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Forest Bear (4)], [Runeclaw Bear (5)])
0: [929] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Runeclaw Bear (5)], [Runeclaw Bear (5)])
0: [879] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Forest Bear (4)], [Forest Bear (4)], [Forest Bear (4)])
0: [1004] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Forest Bear (4)], [Runeclaw Bear (5)], [Grizzly Bears (3)])
0: [929] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Forest Bear (4)], [Runeclaw Bear (5)], [Runeclaw Bear (5)])
0: [929] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Runeclaw Bear (5)], [Grizzly Bears (3)], [Grizzly Bears (3)])
0: [954] Incremental Growth (437) -> Put a +1/+1 counter on target creature, ... (targets: [Runeclaw Bear (5)], [Runeclaw Bear (5)], [Forest Bear (4)])
```


After:
```
0: [1004] Incremental Growth (562) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Forest Bear (4)], [Runeclaw Bear (5)])
0: [929] Incremental Growth (562) -> Put a +1/+1 counter on target creature, ... (targets: [Grizzly Bears (3)], [Runeclaw Bear (5)], [Runeclaw Bear (5)])
0: [879] Incremental Growth (562) -> Put a +1/+1 counter on target creature, ... (targets: [Forest Bear (4)], [Forest Bear (4)], [Forest Bear (4)])
0: [954] Incremental Growth (562) -> Put a +1/+1 counter on target creature, ... (targets: [Forest Bear (4)], [Forest Bear (4)], [Runeclaw Bear (5)])
0: [1004] Incremental Growth (562) -> Put a +1/+1 counter on target creature, ... (targets: [Forest Bear (4)], [Runeclaw Bear (5)], [Grizzly Bears (3)])
0: [929] Incremental Growth (562) -> Put a +1/+1 counter on target creature, ... (targets: [Runeclaw Bear (5)], [Grizzly Bears (3)], [Grizzly Bears (3)])
0: [929] Incremental Growth (562) -> Put a +1/+1 counter on target creature, ... (targets: [Runeclaw Bear (5)], [Forest Bear (4)], [Forest Bear (4)])
```

Not sure why some invalid are still being run though.
Also for some reason it doesn't play it before combat, but that might be another problem.

Are you able to investigate further with this?

Reply:
Thanks for looking into this, just wanted to reply that I saw the message and am planning to investigate more, but it will be a bit before I can get to it. We can keep this open in the meanwhile.

Reply:
no problem, a good fix sometimes takes time, thanks :)

Reply:
Thanks for persuading me to look more into this. I've figured out the problem - the logic for iterating through targets wasn't always clearing the candidate lists whenever it selected a target. I've re-written the algorithm now (and added comments) that properly does this.

I've also verified with your example that no illegal target combinations are simulated. In my test, I've also added a check that the expected number of simulations is being done based on all possible legal combinations.

As for why the AI doesn't do it pre-combat, I think that's an issue elsewhere in the code, as there's logic to decide when to cast spells pre and post combat.

Reply:
Thanks, your logic & comments look reasonable.

However it appears you introduced an infinite loop when lacking enough targets. :/

The behaviour in `selectNextTargets()` with `currentIndex` seems responsible.

Here's my state for testing:
[simtest.txt](https://github.com/Card-Forge/forge/files/10372422/simtest.txt)

Hope you can find it, looking forward to landing this :)

Whereas on master it actually crashes instead:
`Game-0 > java.lang.IllegalArgumentException: Invalid targetIndex=-1
	at forge.ai.simulation.PossibleTargetSelector$Targets.<init>(PossibleTargetSelector.java:40)
	at forge.ai.simulation.PossibleTargetSelector$Targets.<init>(PossibleTargetSelector.java:33)
	at forge.ai.simulation.PossibleTargetSelector.getLastSelectedTargets(PossibleTargetSelector.java:193)
	at forge.ai.simulation.MultiTargetSelector.getLastSelectedTargets(MultiTargetSelector.java:56)
	at forge.ai.simulation.SpellAbilityChoicesIterator.chooseTargets(SpellAbilityChoicesIterator.java:156)
	at forge.ai.simulation.GameSimulator.lambda$0(GameSimulator.java:204)
	at forge.ai.ComputerUtil.handlePlayingSpellAbility(ComputerUtil.java:139)
	at forge.ai.simulation.GameSimulator.simulateSpellAbility(GameSimulator.java:201)
	at forge.ai.simulation.GameSimulator.simulateSpellAbility(GameSimulator.java:154)
	at forge.ai.simulation.SpellAbilityPicker.evaluateSa(SpellAbilityPicker.java:384)
	at forge.ai.simulation.SpellAbilityPicker.chooseSpellAbilityToPlayImpl(SpellAbilityPicker.java:203)`

Reply:
Thanks again! Fixed and added a testcase. PTAL.